qiskit/transpiler/passes/layout/disjoint_utils.py

```diff
@@ -0,0 +1,152 @@
+            # swap cost
+            if dag.num_qubits() <= free_qubits[cmap_index]:
+                out_mapping[cmap_index].append(dag_index)
+                free_qubits[cmap_index] -= dag.num_qubits()
+                break
+        else:
+            raise TranspilerError(
+                "A connected component of the DAGCircuit is too large for any of the connected "
+                "components in the coupling map."
+            )
+    return out_mapping
+
+
+def split_barriers(dag: DAGCircuit):
+    """Mutate an input dag to split barriers into single qubit barriers."""
+    for node in dag.op_nodes(Barrier):
+        num_qubits = len(node.qargs)
+        if num_qubits == 1:
+            continue
+        barrier_uuid = uuid.uuid4()
+        split_dag = DAGCircuit()
+        split_dag.add_qubits([Qubit() for _ in range(num_qubits)])
+        for i in range(num_qubits):
+            split_dag.apply_operation_back(
+                Barrier(1, label=barrier_uuid), qargs=[split_dag.qubits[i]]
+            )
+        dag.substitute_node_with_dag(node, split_dag)
+
+
+def combine_barriers(dag: DAGCircuit, retain_uuid: bool = True):
+    """Mutate input dag to combine barriers with UUID labels into a single barrier."""
+    qubit_indices = {bit: index for index, bit in enumerate(dag.qubits)}
+    uuid_map = {}
+    for node in dag.op_nodes(Barrier):
+        if isinstance(node.op.label, uuid.UUID):
+            barrier_uuid = node.op.label
+            if barrier_uuid in uuid_map:
+                other_node = uuid_map[node.op.label]
+                num_qubits = len(other_node.qargs) + len(node.qargs)
+                new_op = Barrier(num_qubits, label=barrier_uuid)
+                new_node = dag.replace_block_with_op([node, other_node], new_op, qubit_indices)
+                uuid_map[barrier_uuid] = new_node
+            else:
+                uuid_map[barrier_uuid] = node
+    if not retain_uuid:
+        for node in dag.op_nodes(Barrier):
+            if isinstance(node.op.label, uuid.UUID):
+                node.op.label = None
+
+
+def separate_dag(dag: DAGCircuit) -> List[DAGCircuit]:
+    """Separate a dag circuit into it's connected components."""
+    # Split barriers into single qubit barrieries before connected components
+    split_barriers(dag)
+    im_graph, _, qubit_map, __ = vf2_utils.build_interaction_graph(dag)
+    connected_components = rx.weakly_connected_components(im_graph)
+    component_qubits = []
+    for component in connected_components:
+        component_qubits.append(set(qubit_map[x] for x in component))
+
+    qubits = set(dag.qubits)
+
+    decomposed_dags = []
+    for component_qubits in component_qubits:
+        new_dag = dag.copy_empty_like()
+        new_dag.remove_qubits(*qubits - component_qubits)
+        new_dag.global_phase = 0
+        for node in dag.topological_op_nodes():
+            if node.qargs[0] in component_qubits:
```

Comment:
Fixed in: https://github.com/Qiskit/qiskit-terra/pull/9802/commits/70852889950a087ff64ec9ae45605b6a835d186a `node.qargs` is a tuple so I had to reverse it to be a `component_qubits.issuperset(node.qargs)` because the RHS is the only thing that takes an iterable.